[WIP] Improvements in valgrind requests

Varun-garg opened the pull request:
This is extension to Pull Request: #1311 

**Goals**
1. Run `VALGRIND_CHECK_MEM_IS_DEFINED` while freeing objects. (Done)
2. Run Valgrind on Travis.
3. Implement redzones while creating objects.
4. Mark the regions freed as noaccess (Done)
5. Moved from `stl::set` to  `J9HashTable` to remove dependency on libstdc++ (Done)


Signed-off-by: Varun Garg <varun.10@live.com>

0xdaryl commented:
Could you please rebase just your changes on top of HEAD to minimize the number of commits?  When the time comes it will be much easier to review and a much cleaner merge.  Thanks.

Varun-garg commented:
Don't worry @0xdaryl I will take care of commits and rebase everything when I think code for this PR is satisfactory, otherwise I will just have to keep catching up with the HEAD. Goals 1 and 4 have been accomplished, if anyone is interested in reviewing them, I will clean it up now.

Varun-garg commented:
Hey @0xdaryl I just cleaned up commits and rebased at HEAD. Can anyone review the code?

mstoodle commented:
@youngar @Varun-garg is there a reason this work can't be brought into the code base in small pieces, rather than 8 months of lingering WIP pull request?  I'm very happy for this work to happen; I just wonder if we can bring it in more incrementally, or is there a strong reason to go all-or-nothing.

Varun-garg commented:
@mstoodle Thank you for revisiting this PR. It surely can be merged  in pieces (except the last commit). 
I added some code in other jdk related repositories too.

#### Changes:
https://github.com/ibmruntimes/openj9-openjdk-jdk9/compare/openj9...Varun-garg:openj9
https://github.com/eclipse/openj9/compare/master...Varun-garg:master

I have fixed compilation problems and fixed some valgrind warnings in openj9 repo to deal with false positive errors caught by valgrind. It works fine with interpreter mode (`-Djava.compiler=NONE`)  until the last commit. In last commit I enabled valgrind checks for uninitialized values which break some functionality on scavanger. (I can fix it,  meanwhile `-Xgcpolicy:optthruput` option works fine).

#### Currently I am working on:
1. Detect [uninitialized variables in java](https://stackoverflow.com/a/15165301/3276830). I am stuck here for some time. Could someone oint to the code in openjdk where default initialization is happening or where to look for it? Then by letting valgrind know that value is uninitialized we could produce useful errors for java users. *I couldn't find same funtionality elsewhere.*
2. Test how well the integration is working. The PR pushed in GSoC last summer had some mistakes and yet wasn't throwing any errors. Maybe I can add a test case with actual leaks?

mstoodle commented:
Thanks for the additional information. Unfortunately, the ibmruntimes/openj9-openjdk-jdkN repositories are not able to accept outside contributions at the moment. To accept contributions there would mean implementing a contributor agreement along the same lines as the OCA that OpenJDK contributors must sign, and we simply haven't set one up because we didn't expect many contributions there. I haven't reviewed those changes, but are they the kinds of things that could maybe be contributed to OpenJDK directly rather than the OpenJ9 pieces? Or could they be relocated into the Eclipse OpenJ9 project?

What do you mean by uninitialized "variables" ... If you mean locals, then I think javac does a good job already pointing those out. If you're talking about object fields, then that's going to be attached to the process of allocating/constructing an object.

We welcome additional tests, provided they execute in reasonable time. But OMR currently can't take, say, a test written in Java because we don't have a JVM here to run Java code. A Java test would be better created at either Eclipse OpenJ9 or (even better) in the test related repos at AdoptOpenJDK. If it's a Java test you're thinking about, I would reach out to @smlambert to discuss your options.

Varun-garg commented:
Hi @mstoodle, I would try to relocate the cheanges in OpenJDK repository to OpenJ9. Can I directly set flags in OpenJ9 and OMR from the 'configure' script in the root of OpenJDK?

I can demonstrate use for checking uninitalized variables by following program. I can't remember its source, but it compiles without any warning on `javac` (and `-Werror`)

````java
class HelloWorld {

    int x;

    {
        printX();
        x = 7;
        printX();
    }

    HelloWorld() {
        System.out.println("Constructor");
    }

    void printX() {

	if(x == 0)
		System.out.println("Statement 1");
	else
		System.out.println("Statement 2");

    }

    public static void main(String[] args) {
        HelloWorld t = new HelloWorld();
    }
}
````
**Produces**
````
Statement 1
Statement 2
Constructor
````

I was initially thinking about making tests for OMR. But since Java integration is almost finished , testing on Java it could be more useful.

In this PR, I have just removed the last commit regarding uninintialized values. So, from my side it is ready for merging. You can still check it on this tree: https://github.com/Varun-garg/omr/commits/masterBeta

mstoodle commented:
Thanks for the explanation, @Varun-garg . @youngar, can you please review the remaining commits?

> Can I directly set flags in OpenJ9 and OMR from the 'configure' script in the root of OpenJDK

Alas, I suspect OpenJDK probably won't like that (though, you could always try...it would be interesting to see their response :) ). Is there any way the changes you want could be integrated into the OpenJ9 build system? For the record, this really isn't an OMR discussion; we should really take it to the OpenJ9 project.

mstoodle commented:
@youngar do you have any concerns to merge this PR ?

youngar commented:
Hi, I think I forgot to hit the submit review (with comments) months ago, sorry about that.  I had assumed that interest in finishing the work was lost! Sorry about that, I'll start a review from scratch.

Varun-garg commented:
Thanks for reviewing this work @youngar . I will push the requested changes shortly.

youngar commented:
@Varun-garg Hi Varun, I think there are probably more changes needed for this code, i'll do another review once you push.

I'm sorry you couldn't contact me.  You should be able to find me on the [OpenJ9 slack](https://www.eclipse.org/openj9/oj9_joinslack.html).

Varun-garg commented on gc/base/MemcheckWrapper.cpp:
```diff
@@ -27,46 +27,108 @@
+	const uintptr_t* data;
+	uintptr_t dataLength;
+	BOOLEAN forceCollisions;
+} HashtableInputData;
```
Seems like this change was already done.

Varun-garg commented on gc/base/MemcheckWrapper.cpp:
```diff
@@ -27,46 +27,108 @@
+static uintptr_t hashFn(void *key, void *userData)
+{
+	BOOLEAN forceCollisions = (BOOLEAN)((uintptr_t)userData);
+	if (forceCollisions) {
```
This one too.

Varun-garg commented on gc/base/MemcheckWrapper.cpp:
```diff
@@ -75,67 +137,99 @@ void valgrindMakeMemNoaccess(uintptr_t address, uintptr_t size)
+            valgrindFreeObjectDirect(extensions,*currentSlotPointer);
+            it.removeSlot();
         }
+        currentSlotPointer = (uintptr_t*)it.nextSlot(); // TODO: 
```
It looks fine, maybe I was thinking of something else back then and should have at completed the line.

Varun-garg commented on gc/base/MemcheckWrapper.cpp:
```diff
@@ -75,67 +137,99 @@ void valgrindMakeMemNoaccess(uintptr_t address, uintptr_t size)
 
 void valgrindFreeObject(MM_GCExtensionsBase *extensions, uintptr_t baseAddress)
 {
+    int objSize;
```
Problem with this approach is that now valgrindFreeObjectDirect doesn't handle the objects that have become forwarded pointer but valgrindFreeObject does.

Varun-garg commented on gc/base/GCExtensionsBase.hpp:
```diff
@@ -736,8 +732,8 @@ class MM_GCExtensionsBase : public MM_BaseVirtual {
 #if defined(OMR_VALGRIND_MEMCHECK)
-	uintptr_t valgrindMempoolAddr; /** <Memory pool's address for valgrind> **/
-	std::set<uintptr_t> _allocatedObjects;
+	uintptr_t valgrindMempoolAddr; /** < Memory pool's address for valgrind **/
```
Fixed.

Varun-garg commented on gc/base/GCExtensionsBase.hpp:
```diff
@@ -736,8 +732,8 @@ class MM_GCExtensionsBase : public MM_BaseVirtual {
-	uintptr_t valgrindMempoolAddr; /** <Memory pool's address for valgrind> **/
-	std::set<uintptr_t> _allocatedObjects;
+	uintptr_t valgrindMempoolAddr; /** < Memory pool's address for valgrind **/
+	J9HashTable *MemcheckHashTable = NULL; /** <Hash table to store object addresses for valgrind> **/
```
Fixed.

Varun-garg commented on gc/base/MemcheckWrapper.cpp:
```diff
@@ -27,46 +27,108 @@
+	}
+}
+
+static J9HashTable * allocateHashtable(OMRPortLibrary *portLib, HashtableInputData *inputData)
```
Fixed.

Varun-garg commented on gc/base/MemcheckWrapper.cpp:
```diff
@@ -27,46 +27,108 @@
+			entrySize,
+			sizeof(char *),
+			flags | J9HASH_TABLE_ALLOW_SIZE_OPTIMIZATION,
+			OMRMEM_CATEGORY_VM,
```
Fixed.

Varun-garg commented on gc/base/MemcheckWrapper.cpp:
```diff
@@ -27,46 +27,108 @@
+			hashFn,
+			hashEqualFn,
+			NULL,
+			userData);
```
Already done.

Varun-garg commented on gc/base/MemcheckWrapper.cpp:
```diff
@@ -75,67 +137,99 @@ void valgrindMakeMemNoaccess(uintptr_t address, uintptr_t size)
-            VALGRIND_PRINTF("Erasing set from lBound = %lx to uBound = %lx\n",*lBound,*uBound);		
-#endif /* defined(VALGRIND_REQUEST_LOGS) */			
-            extensions->_allocatedObjects.erase(lBound,++uBound);//uBound is exclusive
+            valgrindFreeObjectDirect(extensions,*currentSlotPointer);
```
Fixed.

Varun-garg commented on gc/base/MemcheckWrapper.cpp:
```diff
@@ -75,67 +137,99 @@ void valgrindMakeMemNoaccess(uintptr_t address, uintptr_t size)
+
+    /* Valgrind automatically marks free objects as noaccess.
+    We still mark the entire region as no access for any left out areas */
+    valgrindMakeMemNoaccess(baseAddress,size);
```
Fixed.

Varun-garg commented on gc/base/MemcheckWrapper.cpp:
```diff
@@ -75,67 +137,99 @@ void valgrindMakeMemNoaccess(uintptr_t address, uintptr_t size)
+    VALGRIND_PRINTF("Checking for object at 0x%lx\n", baseAddress);
+#endif /* defined(VALGRIND_REQUEST_LOGS) */
+
+    if(hashTableFind(extensions->MemcheckHashTable,&baseAddress) != NULL)
```
Fixed.

Varun-garg commented on gc/base/MemcheckWrapper.cpp:
```diff
@@ -29,113 +29,207 @@
+	return hashtable;
+}
+
+MMINLINE void valgrindFreeObjectDirect(MM_GCExtensionsBase *extensions, uintptr_t baseAddress);
```
Fixed.

Varun-garg commented on gc/base/MemcheckWrapper.cpp:
```diff
@@ -29,113 +29,207 @@
 #include "GCExtensionsBase.hpp"
+#include "EnvironmentBase.hpp"
+#include "hashtable_api.h"
+#include "HashTableIterator.hpp"
```
Fixed.

Varun-garg commented on gc/base/GCExtensionsBase.hpp:
```diff
@@ -735,8 +731,9 @@ class MM_GCExtensionsBase : public MM_BaseVirtual {
-	std::set<uintptr_t> _allocatedObjects;
+	uintptr_t valgrindMempoolAddr; /** < Memory pool's address for valgrind **/
+	J9HashTable *MemcheckHashTable; /** <Hash table to store object addresses for valgrind> **/
+	MUTEX MemcheckHashTable_mutex;
```
Fixed.

Varun-garg commented:
@youngar I have made the changes you posted.  Please check them and let me know of the other changes.

youngar commented on gc/base/MemorySubSpaceSemiSpace.cpp:
```diff
@@ -607,7 +607,7 @@ MM_MemorySubSpaceSemiSpace::poisonEvacuateSpace()
 	}
 
-#if defined(OMR_VALGRIND_MEMCHECK)			
+#if defined(OMR_VALGRIND_MEMCHECK)
```
Have to update file copyright from 2017

youngar commented:
Besides the one missing copyright date, these changes look good.  I like how you simplified the TLH by marking the memory undefined, it really simplified the code.
@genie-omr build all

Varun-garg commented:
Thank you @youngar . I have fixed copyright directly from Github. Should I also squash it in last commit?

Varun-garg commented:
Accidentally closed.

youngar commented:
Its okay, thanks for fixing it!

mstoodle commented:
Well post this PR having been merged, I removed the `[WIP]` marker in the title since obviously it wasn't any more. In future, please remove / ask for the `[WIP]` to be removed before merging. Thanks @youngar @Varun-garg !
